lib/puppet/rest/client.rb

```diff
@@ -0,0 +1,83 @@
+require 'httpclient'
+
+require 'puppet'
+require 'puppet/rest/response'
+
+module Puppet::Rest
+  class Client
+    # Create a new HTTP client for querying the given API.
+    # @param [Puppet::Rest::Route] route data about the API being queried,
+    #                              including the API name and server details.
+    # @param [OpenSSL::X509::Store] ssl_store the SSL configuration for this client
+    # @param [Integer] receive_timeout how long in seconds this client will wait
+    #                  for a response after making a request
+    # @param [HTTPClient] client the third-party HTTP client wrapped by this
+    #                     class. This param is only used for testing.
+    def initialize(route,
+                   ssl_store: OpenSSL::X509::Store.new,
+                   receive_timeout: Puppet[:http_read_timeout],
+                   client: HTTPClient.new(agent_name: nil,
+                                          default_header: {
+                                            'User-Agent' => Puppet[:http_user_agent],
+                                            'X-PUPPET-VERSION' => Puppet::PUPPETVERSION
+                                          }))
+      @client = client
+
+      configure_client(ssl_store, route, receive_timeout)
+    end
+
+    # Configures the underlying HTTPClient
+    # @param [OpenSSL::X509::Store] ssl_store the SSL configuration for this client
+    # @param [Puppet::Rest::Route] route data about the API being queried
+    # @param [Integer] timeout how long to wait for a response from the server once
+    #                  a request has been made
+    def configure_client(ssl_store, route, timeout)
```

Comment:
I didn't because these are private methods that can't (or shouldn't) be called outside the context of this class that has `client` as a private member. But I don't have a strong opinion.